bmi_sdk/bmi_sdk/bmi_variable.py

```diff
@@ -0,0 +1,128 @@
+from __future__ import annotations
+
+from enum import Enum
+
+# Once python 3.8 is no longer supported, just get Annotated directly
+# from typing import Annotated, Any, Literal, Optional, Self, Union
+from typing import Any, Literal, Optional, Union
+
+import numpy as np
+from pydantic import (
+    BaseModel,
+    ConfigDict,
+    Field,
+    RootModel,
+    computed_field,
+    model_validator,
+)
+
+# Only needed on python 3.8
+from typing_extensions import Annotated, Self
+
+from .bmi_grid import Grid
+
+Integer = RootModel[Annotated[str, Literal["int", "int16", "int32", "integer"]]]
+Long = RootModel[Annotated[str, Literal["int", "int64", "longlong"]]]
+Float = RootModel[
+    Annotated[
+        str,
+        Literal[
+            "float",
+            "float32",
+            "np.float32",
+            "numpy.float32",
+            "np.single",
+            "numpy.single",
+        ],
+    ]
+]
+Double = RootModel[
+    Annotated[str, Literal["float", "float64", "np.float64", "numpy.float64"]]
+]
+
+VarType = Union[Integer, Long, Float, Double]
+
+
+class VarLocation(str, Enum):
+    """Enumeration of variable grid location strings supported by BMI.
+
+    See https://bmi.readthedocs.io/en/stable/#get-var-location
+
+    """
+
+    node = "node"
+    edge = "edge"
+    face = "face"
+
+
+class BmiVariable(BaseModel):
```

Comment:
Just `Variable`

Reply:
This one I'm not sold on changing, but would consider other names.  I feel like Variable by itself may loose context in code, e.g.

```python
from bmi_sdk.bmi_variable import Variable
...
...
...
#many lines of code later
something = Variable()
```
I'll have to weigh that cognitive effort a bit.

Reply:
I agree with Nels here. It's like naming something`np_array`, the contents are not totally obvious.

Reply:
@hellkite500, yeah that is fair. For better or worse that is the predominant way imports are made in python. I think the way `pandas` and `numpy` advise importing (e.g. `import numpy as np`) is much more clear, but then again that is just dogma.

Reply:
That said, nothing stops you from doing something like this
```python
from bmi_sdk.bmi_variable import BmiVariable as bv
...
something = bv()
```
Python makes the aliasing too flexible to try to utilize the namespace in a contextually consistent way, haha.  So being a little verbose in some cases doesn't hurt.  We could also provide module/package aliases in the future if we wanted to.

Reply:
Yeah, that is fair. `bmi_sdk` is a little long to use everywhere anyways. We could in all examples do, `import bmi_sdk as bk` and then you would just do `bk.Variable` (if we export at the package level). That or we could reconsider the name of the package to something like `bmisk` instead?? Just general thoughts. Just feels very C like to have to say where the object came from in its name.

> Python makes the aliasing too flexible to try to utilize the namespace in a contextually consistent way, haha. So being a little verbose in some cases doesn't hurt.

Yeah, I think a lot of this comes down to how you document usage of the library. So, if in all the examples you do `import pandas as pd` and encourage that, then I think people will generally follow that. Of course we are no `pandas`, but `pandas`, `numpy`, and `seaborn` are all examples that come to mind.